### PR TITLE
CT-2049: Add CommandCompleted event

### LINKED ISSUE #6878: [CT-2049] Fire an event at the end of every command
On the `feature/click-cli` branch, each command explicitly returns a tuple of `success: bool, results: Any`. (The structure of `results` differs based on the command.)

We should additionally fire an event that can be used programmatically to determine:
- when the command completed
- if the command succeeded

Ideally, this event would be fired _always_, even in cases where the command was interrupted (`^C`) or cancelled itself (e.g. `--fail-fast`).

### PR DESCRIPTION
**Pull request by QMalcolm**
resolves #6878 

### Description

[CT-2049](https://github.com/dbt-labs/dbt-core/issues/6878) outlines a desire to have an event that is output on completion of any and every command. This PR adds that functionality. Below is a demo of the functionality in a loom video (clicking will redirect you to loom)

<a href="https://www.loom.com/share/661cb790b6524c4b87ecd90ca3520f00">
    
</a>

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

[CT-2049]: https://dbtlabs.atlassian.net/browse/CT-2049?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/dbt/events/types.py [hunk 1]
```diff
@@ -1827,6 +1827,16 @@ def message(self) -> str:
         return "No nodes selected!"
 
 
+@dataclass
+class CommandCompleted(InfoLevel, pt.CommandCompleted):
```
Comment (jtcohen6) [inline core/dbt/events/types.py:1831]: This feels more `DebugLevel` than `InfoLevel` to me - do you feel strongly that it should be in the stdout logs?

File: .changes/unreleased/Under the Hood-20230316-123305.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+kind: Under the Hood
+body: Add CommandComplete event, and fire it upon completion of every command
```
Comment (jtcohen6) [inline .changes/unreleased/Under the Hood-20230316-123305.yaml:2]: Nit: `CommandCompleted`

**Review by jtcohen6 — COMMENTED** [review]
yo!! nice quick work - love to see an early draft PR :)

File: core/dbt/events/types.py (hunk 1 above)
Comment (QMalcolm) [reply core/dbt/events/types.py:1831]: I have no strong opinion. I'm fine with it being a debug 👍

File: .changes/unreleased/Under the Hood-20230316-123305.yaml (hunk 2 above)
Comment (QMalcolm) [reply .changes/unreleased/Under the Hood-20230316-123305.yaml]: I corrected the typo in a fixup commit. Changie entry now has commit sha https://github.com/dbt-labs/dbt-core/pull/7180/commits/9860e1130da9db2684c48b6ad4b9a8f363922998

File: core/dbt/events/types.py (hunk 1 above)
Comment (QMalcolm) [reply core/dbt/events/types.py]: Updated to a `Debug` level in https://github.com/dbt-labs/dbt-core/pull/7180/commits/92cec272ca70c276b892bdec412f46a73fe15ec0

File: core/dbt/cli/requires.py [hunk 3]
```diff
@@ -53,7 +60,33 @@ def wrapper(*args, **kwargs):
...
+                    command=ctx.command_path,
+                    success=success,
+                    completed_at=datetime.datetime.utcnow(),
+                    elapsed=time.perf_counter() - start_func,
+                )
+            )
+        # Bare except because we really do want to catch ALL exceptions,
+        # i.e. we want to fire this event in ALL cases.
+        except:  # noqa
+            fire_event(
```
[the remark this answers is not in the document]
Comment (ChenyuLInx) [reply core/dbt/cli/requires.py:79]: Totally okay doing it separately to keep each PR smaller.

[the remark this answers is not in the document]
Comment (QMalcolm) [reply core/dbt/cli/requires.py:79]: I agree that we should definitely work to fire that event again (and handle the fail fast exceptions better). I'm not opposed to taking on that work. However I'd personally prefer to make it a separate PR to keep the PR process as smooth as possible.

[the remark this answers is not in the document]
Comment (jtcohen6) [reply core/dbt/cli/requires.py:79]: Agree - let's do it in a separate PR! Even though that issue is not scheduled for the current sprint, we definitely want it before the v1.5 RC (tracked in https://github.com/dbt-labs/dbt-core/issues/7162)

**Review by gshank — APPROVED** [review]
Looks good.

**Review by stu-k — COMMENTED** [review]
I know it's already a little bloated, but it might be nice to move your work out of the `preflight` decorator as that's meant to contain setup type initialization, not managing the resulting data.

**Comment by QMalcolm** [comment]
> I know it's already a little bloated, but it might be nice to move your work out of the `preflight` decorator as that's meant to contain setup type initialization, not managing the resulting data.

Hey @stu-k! I'm happy to move it out of the `preflight` decorator 😄 The only question is, "where?" 🤔 It's worth mentioning that the "where" was [previously discussed and the conclusion at the time was that the preflight decorator might be the best place](https://github.com/dbt-labs/dbt-core/issues/6878#issuecomment-1428643539). Looking at the existing decorators, I don't believe there is another good candidate currently. Thus the alternative would be to create a new decorator simply for the handling of this functionality, and using it with every command. Let me know what you think 🙂

**Review by ChenyuLInx — APPROVED** [review]
LGTM!!! Thanks for also have a loom videos showing how this works under all of the failure modes!

**Comment by ChenyuLInx** [comment]
Also agreed on let's leave the `move it out of the preflight decorator` out of this PR. Since it probably apply to both this issue and the exit code issue.

**Comment by stu-k** [comment]
@QMalcolm We can follow up on the moving into another place in another ticket!

**Pull request merged**